File: prisoned-in-c/formulas.c
#include <math.h>
#include <stdlib.h>
#include "game.h"
#include <stdio.h>
#include "parameters.h"


float fermi(float point1, float point2, ParamPtr param) {
    float prob = 1 / (1 + exp((point1 - point2) / param->k));
    return prob;
}

float H_x(int memory, ParamPtr param) {
    if (param->max_memory == 0){
        return 1;
    }else{
        return 1-(1-param->beta)*memory/param->max_memory;}
}

float oneGame_oneResult(Cell_ptr player1, Cell_ptr player2, ParamPtr param){
    char move1 = player1->strategy();
    char move2 = player2->strategy();

    if ((move1 == 'C') && (move2 == 'C')) {
        return param->r;
    }
    else if ((move1 == 'C') && (move2 == 'D')) {
        return param->s;
    }
    else if ((move1 == 'D') && (move2 == 'C')) {
        return param->t;
    }
    else if ((move1 == 'D') && (move2 == 'D')) {
        return param->p;
    }
    else {
        return -1;
    }
}

int incrementPoint(Cell_ptr player1, Cell_ptr player2, ParamPtr param){
    player1->point += oneGame_oneResult(player1, player2, param);
    return 0;
}

int changeStrategy(Cell_ptr player1, Cell_ptr player2, ParamPtr param){
    float thisrand = rand();
//    printf("maybe change");  //to debug
    
    float prob = H_x(player1->memory, param)*fermi(player1->point, player2->point, param);
    char (*newstrategy)() = player1->strategy;
    
    thisrand /= RAND_MAX;
//    printf(" %f, %f, %i \n", prob, thisrand, myrand); //to debugr
    if (prob > thisrand){
//        printf("changing strategy\n"); //to debug
        newstrategy = player2->strategy;
    }
    if (newstrategy == player1->strategy){
        if (player1->memory<param->max_memory){
            player1->memory++;
        }
    }
    else{
        player1->memory = 0;
        player1->strategy = newstrategy;
    }
    return 0;
}

File: prisoned-in-julia/formulas.jl
include("game.jl")
#include("iperparametri.jl")

#K = iperparametri.K
#Max_memory = iperparametri.Max_memory
#Rand_Max = iperparametri.Rand_Max

function fermi(point1::Float64, point2::Float64)::Float64
    prob = 1 / (1 + exp((point1 - point2) / K))
    return prob
end

function H_x(memory::Int)::Float64
    if Max_memory == 0
        return 1
    else
        return 1-(1-beta)*memory/Max_memory
    end
end

function incrementPoint(player1::Cell, player2::Cell)::Int
    player1.point += oneGame_oneResult(player1, player2)
    return 0
end

function changeStrategy(player1::Cell, player2::Cell)
    thisrand = rand()/Rand_Max
    prob = H_x(player1.memory)*fermi(player1.point, player2.point)
    newstrategy = player1.strategy
    if prob<thisrand
        newstrategy = player2.strategy
    end
    if newstrategy == player1.strategy
        if player1.memory < Max_memory
            player1.memory +=1
        end
    else
        player1.memory = 0
        player1.strategy = newstrategy
    end
end